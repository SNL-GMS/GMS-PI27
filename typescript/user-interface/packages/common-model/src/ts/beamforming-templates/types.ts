import type { CommonTypes, EventTypes, SignalDetectionTypes } from '../common-model';
import type { EntityReference, VersionReference } from '../faceted';
import type { FilterDefinition } from '../filter';
import type {
  Channel,
  OrientationAngles
} from '../station-definitions/channel-definitions/channel-definitions';
import type { Station } from '../station-definitions/station-definitions/station-definitions';

export enum BeamType {
  CONTINUOUS_LOCATION = 'CONTINUOUS_LOCATION',
  DETECTION = 'DETECTION',
  EVENT = 'EVENT',
  FK = 'FK'
}

// TODO: beamSummationType enum instead of string
export interface BeamDescription {
  readonly beamSummation: string;
  readonly beamType: BeamType;
  readonly phase: string;
  readonly samplingType: string;
  readonly twoDimensional: boolean;
  readonly preFilterDefinition?: FilterDefinition;
}

export interface BeamformingTemplate {
  readonly beamDescription: BeamDescription;
  readonly inputChannels: VersionReference<'name'>[] | Channel[];
  readonly minWaveformsToBeam: number;
  readonly orientationAngleToleranceDeg: number;
  readonly sampleRateToleranceHz: number;
  readonly station: VersionReference<'name'> | Station;
  readonly beamDuration?: number;
  readonly leadDuration?: number;
}

export interface BeamParameters {
  eventHypothesis?: EntityReference<'id', EventTypes.EventHypothesis>;
  location?: CommonTypes.Location;
  minWaveformsToBeam: number;
  orientationAngles: OrientationAngles;
  orientationAngleToleranceDeg: number;
  receiverToSourceAzimuthDeg: number | undefined;
  sampleRateHz: number;
  sampleRateToleranceHz: number;
  signalDetectionHypothesis?: SignalDetectionTypes.SignalDetectionHypothesis;
  slownessSecPerDeg: number | undefined;
}

export interface BeamDefinition {
  beamDescription: BeamDescription;
  beamParameters: BeamParameters;
}

export type BeamformingTemplatesByPhase = Record<string, BeamformingTemplate>;

export type BeamformingTemplatesByStationByPhase = Record<string, BeamformingTemplatesByPhase>;

export type BeamformingTemplatesByBeamTypeByStationByPhase = Record<
  string,
  BeamformingTemplatesByStationByPhase
>;
