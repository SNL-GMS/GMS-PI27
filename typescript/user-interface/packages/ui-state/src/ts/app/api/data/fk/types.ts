/**
 * Input type for Compute FK service call. This input
 * is compatible with the COI input i.e. start/end are strings
 */
import type {
  ChannelTypes,
  FkTypes,
  ProcessingMaskDefinitionTypes,
  StationTypes,
  WaveformTypes
} from '@gms/common-model';

import type { UiChannelSegment } from '../../../../types';
import type { AsyncFetchHistory } from '../../../query';

/**
 * Defines the history record type for the getChannelSegmentsByChannel query
 */
export type ComputeFkSpectraHistory = AsyncFetchHistory<FkTypes.FkInputWithConfiguration>;

/**
 * Arguments required for the ComputeFkSpectra operation
 */
export interface ComputeFkSpectraArgs {
  readonly fkSpectraDefinition: FkTypes.FkSpectraDefinition;
  readonly station: StationTypes.Station;
  readonly inputChannels: ChannelTypes.Channel[];
  readonly detectionTime: number;
  readonly startTime: number;
  readonly endTime: number;
  readonly uiChannelSegments: UiChannelSegment<WaveformTypes.Waveform>[];
  readonly processingMasksByChannel: FkTypes.ProcessingMasksByChannel[];
  readonly maskTaperDefinition: ProcessingMaskDefinitionTypes.TaperDefinition | undefined;
  readonly expandedTimeBufferSeconds: number;
}
