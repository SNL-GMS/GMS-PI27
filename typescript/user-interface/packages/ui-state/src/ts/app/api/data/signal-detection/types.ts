import type { SignalDetectionTypes } from '@gms/common-model';

/**
 * Arguments to modify existing Signal Detection
 */
export interface UpdateSignalDetectionArrivalTimeArgs {
  readonly signalDetectionId: SignalDetectionTypes.SignalDetectionIdString;
  readonly arrivalTime: {
    readonly value: number;
    readonly uncertainty: number;
  };
}

/**
 * Arguments to delete a signal detection
 */
export interface DeleteSignalDetectionArgs {
  readonly signalDetectionIds: string[];
}
