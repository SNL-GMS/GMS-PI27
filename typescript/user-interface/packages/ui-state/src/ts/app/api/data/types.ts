import type { BeamformingTemplateTypes, FkTypes, RotationTypes } from '@gms/common-model';

import type {
  AssociationConflictRecord,
  ChannelRecord,
  EventsRecord,
  FilterDefinitionsForSignalDetectionsRecord,
  FilterDefinitionsRecord,
  FkChannelSegmentRecord,
  FkFrequencyThumbnailRecord,
  QcSegmentRecord,
  SignalDetectionsRecord,
  UIChannelSegmentRecord
} from '../../../types';
import type { GetChannelsByNamesHistory } from './channel/types';
import type {
  PredictFeaturesForEventLocation,
  PredictFeaturesForEventLocationHistory
} from './event/predict-features-for-event-location';
import type {
  FindEventsByAssociatedSignalDetectionHypothesesHistory,
  GetEventsWithDetectionsAndSegmentsByTimeHistory
} from './event/types';
import type { ComputeFkSpectraHistory } from './fk/types';
import type { GetFilterDefinitionsForSignalDetectionsHistory } from './signal-detection/get-filter-definitions-for-signal-detections';
import type { GetSignalDetectionsWithSegmentsByStationAndTimeHistory } from './signal-detection/get-signal-detections-segments-by-station-time';
import type {
  FilterDefinitionByUsage,
  FilterDefinitionByUsageRecord,
  GetBeamformingTemplatesHistory,
  GetFilterDefinitionByUsageHistory,
  GetFkReviewablePhasesHistory,
  GetFkSpectraTemplatesHistory,
  GetProcessingMaskDefinitionsHistory,
  GetRotationTemplatesHistory,
  ProcessingMaskDefinitionByProcessingOperationByPhaseByChannel
} from './signal-enhancement';
import type {
  FindEventBeamsByEventHypothesisAndStationsHistory,
  UiChannelSegmentByEventHypothesisId
} from './waveform';
import type { FindQCSegmentsByChannelAndTimeRangeHistory } from './waveform/find-qc-segments-by-channel-and-time-range';
import type { GetChannelSegmentsByChannelHistory } from './waveform/get-channel-segments-by-channel';

/**
 * Defines the Data slice state.
 */
export interface DataState {
  /** the channel segments - by unique channel name - populated by multiple queries */
  uiChannelSegments: UIChannelSegmentRecord;

  /** FK channel segment records */
  fkChannelSegments: FkChannelSegmentRecord;

  /** Fk frequency thumbnails record */
  fkFrequencyThumbnails: FkFrequencyThumbnailRecord;

  /** the channels by effectiveTime populated by the getChannelsByNamesTimeRange query */
  channels: {
    raw: ChannelRecord;
    beamed: ChannelRecord;
    filtered: ChannelRecord;
  };
  /** the signal detections - by signal detection id - populated by multiple queries */
  signalDetections: SignalDetectionsRecord;
  /** the events - by time- populated by multiple queries */
  events: EventsRecord;
  eventBeams: UiChannelSegmentByEventHypothesisId;
  /** the association conflicts for events in conflict */
  associationConflict: AssociationConflictRecord;
  /** designed filter definitions */
  filterDefinitions: FilterDefinitionsRecord;
  /** filter definitions for signal detections */
  filterDefinitionsForSignalDetections: FilterDefinitionsForSignalDetectionsRecord;
  /** default filter definitions */
  defaultFilterDefinitionsMap: FilterDefinitionByUsage;
  /** qc segments by unique channel name */
  qcSegments: QcSegmentRecord;
  /** Processing Mask definitions */
  processingMaskDefinitionsByChannels: ProcessingMaskDefinitionByProcessingOperationByPhaseByChannel[];
  /** Beamforming templates */
  beamformingTemplates: BeamformingTemplateTypes.BeamformingTemplatesByBeamTypeByStationByPhase;
  /** Rotation templates */
  rotationTemplates: RotationTypes.RotationTemplateByPhaseByStationRecord;
  /** filter definitions by usage map */
  filterDefinitionsForUsage: FilterDefinitionByUsageRecord;
  /** feature predictions for event locations */
  predictFeaturesForEventLocation: PredictFeaturesForEventLocation;

  /**
   * FK Reviewable phases, organized as a record of records.
   * Top-level key is an activity name, inner record key is a station name.
   */
  fkReviewablePhases: FkTypes.FkReviewablePhasesByActivityNameByStation;
  /** FK Spectra Templates */
  fkSpectraTemplates: FkTypes.FkSpectraTemplatesByStationByPhase;
  /** query history */
  queries: {
    /** the history record of the computeFkSpectra call */
    computeFkSpectra: ComputeFkSpectraHistory;
    /** the history record of the getChannelSegmentsByChannel query */
    getChannelSegmentsByChannel: GetChannelSegmentsByChannelHistory;
    /** the history record of the findQCSegmentsByChannelAndTimeRange query */
    findQCSegmentsByChannelAndTimeRange: FindQCSegmentsByChannelAndTimeRangeHistory;
    /** the history record of the findEventBeamsByEventHypothesisAndStations query */
    findEventBeamsByEventHypothesisAndStations: FindEventBeamsByEventHypothesisAndStationsHistory;
    /** the history record of the getSignalDetectionWithSegmentsByStationAndTime query */
    getSignalDetectionWithSegmentsByStationAndTime: GetSignalDetectionsWithSegmentsByStationAndTimeHistory;
    /** the history record of the getEventsWithDetectionsAndSegmentsByTime query */
    getEventsWithDetectionsAndSegmentsByTime: GetEventsWithDetectionsAndSegmentsByTimeHistory;
    /** the history record of the findEventsByAssociatedSignalDetectionHypotheses query */
    findEventsByAssociatedSignalDetectionHypotheses: FindEventsByAssociatedSignalDetectionHypothesesHistory;
    /** the history record of the getChannelsByName query */
    getChannelsByNamesTimeRange: GetChannelsByNamesHistory;
    /** the history record of the getFilterDefinitionsByUsage query */
    getFilterDefinitionsForSignalDetections: GetFilterDefinitionsForSignalDetectionsHistory;
    /** the history record of the getProcessingMaskDefinitions query */
    getProcessingMaskDefinitions: GetProcessingMaskDefinitionsHistory;
    /** the history record of the getBeamformingTemplates query */
    getBeamformingTemplates: GetBeamformingTemplatesHistory;
    /** the history record of the getRotationTemplates query */
    getRotationTemplates: GetRotationTemplatesHistory;
    /** the history record of the getFkReviewablePhases query */
    getFkReviewablePhases: GetFkReviewablePhasesHistory;
    /** the history record of the getFkSpectraTemplates query */
    getFkSpectraTemplates: GetFkSpectraTemplatesHistory;
    /** the history record of the getFilterDefinitionsByUsageMap */
    getFilterDefinitionsByUsageMap: GetFilterDefinitionByUsageHistory;
    /** the history record of the {@link PredictFeaturesForEventLocationHistory} */
    predictFeaturesForEventLocation: PredictFeaturesForEventLocationHistory;
  };
}

export type DataStateKeys = keyof DataState;
