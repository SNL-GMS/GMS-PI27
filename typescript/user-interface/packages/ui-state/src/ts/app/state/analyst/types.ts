import type { ChannelSegmentTypes, WaveformTypes } from '@gms/common-model';
import { createEnumTypeGuard } from '@gms/common-util';
import Immutable from 'immutable';

export enum AnalystKeyAction {
  ESCAPE = 'Escape',
  SAVE_OPEN_EVENT = 'Save Open Event',
  SAVE_ALL_EVENTS = 'Save All Events in Interval'
}

export const AnalystKeyActions: Immutable.Map<string, AnalystKeyAction> = Immutable.Map([
  ['Control+KeyS', AnalystKeyAction.SAVE_OPEN_EVENT],
  ['Control+Shift+KeyS', AnalystKeyAction.SAVE_ALL_EVENTS],
  ['Escape', AnalystKeyAction.ESCAPE]
]);

/**
 * The display mode options for the waveform display.
 */
export enum WaveformDisplayMode {
  DEFAULT = 'Default',
  MEASUREMENT = 'Measurement'
}
/**
 * Available waveform align types.
 */
export enum AlignWaveformsOn {
  TIME = 'Time',
  PREDICTED_PHASE = 'Predicted',
  OBSERVED_PHASE = 'Observed'
}

/**
 * Available waveform sort types.
 */
export enum WaveformSortType {
  distance = 'Distance',
  stationNameAZ = 'Station: A-Z',
  stationNameZA = 'Station: Z-A'
}

export type DisplayedMagnitudeTypes = Record<string, boolean>;

export const isAnalystKeyAction = createEnumTypeGuard(AnalystKeyAction);

/**
 * Measurement mode state.
 */
export interface MeasurementMode {
  /** The display mode */
  mode: WaveformDisplayMode;

  /**
   * Measurement entries that are manually added or hidden by the user.
   * The key is the signal detection id
   */
  entries: Record<string, boolean>;
}

/**
 * A list such that at each index, it indicates whether the filter at the
 * same index is within the hotkey cycle.
 */
export type HotkeyCycleList = (boolean | null)[];

/**
 * The location solution state.
 * Includes:
 *   * The selected location solution set and solution
 *   * The selected preferred location solution set and solution
 */
export interface LocationSolutionState {
  selectedLocationSolutionId: string;
  selectedLocationSolutionSetId: string;
  selectedPreferredLocationSolutionId: string;
  selectedPreferredLocationSolutionSetId: string;
}

export interface LoadingActionStatus {
  isComplete?: boolean;
  error?: string;
  actionType: 'REQUEST' | 'CLIENT_SIDE_ACTION';
}

export interface RequestStatus extends LoadingActionStatus {
  url?: string;
  actionType: 'REQUEST';
}

export interface ClientActionStatus extends LoadingActionStatus {
  clientAction?: string;
  actionType: 'CLIENT_SIDE_ACTION';
}

export const isRequestStatus = (statusObj: LoadingActionStatus): statusObj is RequestStatus => {
  return statusObj?.actionType === 'REQUEST';
};

export const isClientActionStatus = (
  statusObj: LoadingActionStatus
): statusObj is ClientActionStatus => {
  return statusObj.actionType === 'CLIENT_SIDE_ACTION';
};

/**
 * Type used for determining the analyst action on an event
 */
export type EventActionTypes = 'duplicate' | 'reject' | 'delete' | 'open' | 'close' | 'details';

/**
 * Type used for determining the analyst action on a signal detection
 */
export type SignalDetectionActionTypes =
  | 'reject associate'
  | 'delete'
  | 'details'
  | 'associate'
  | 'unassociate'
  | 'phase'
  | 'default phase'
  | 'current phase'
  | 'fk'
  | 'rotate';

/**
 * union of action types
 */
export type ActionTypes = EventActionTypes | SignalDetectionActionTypes | null;

export interface ActionTarget {
  actionType: ActionTypes; // the type of action the analyst will take on an event or sd
  eventIds: string[]; // The ids of the events that are the target of a user's action
  signalDetectionIds: string[];
}

export interface AnalystState {
  channelFilters: Record<string, WaveformTypes.WaveformFilter>;
  defaultSignalDetectionPhase: string;
  effectiveNowTime: number;
  fileId: {
    workflowUniqueId: string;
    fileHandle: FileSystemFileHandle | null;
  };
  hotkeyCycleOverrides: Record<string, Record<number, boolean>>;
  location: LocationSolutionState;
  measurementMode: MeasurementMode;
  requestTracker: {
    /* record keyed on ids */
    requests: Record<string, RequestStatus | ClientActionStatus>;
    initiatedRequests: number;
    completedRequests: number;
    lastRequestId: string;
  };
  openEventId: string;
  actionTargets: ActionTarget;
  currentPhase: string;
  selectedEventIds: string[];
  selectedFilterIndex: number | null;
  selectedFilterList: string;
  selectedSdIds: string[];
  selectedSortType: WaveformSortType;

  eventListOpenEventTriggered: boolean;
  mapOpenEventTriggered: boolean;
  alignWaveformsOn: AlignWaveformsOn;
  phaseToAlignOn: string;
  phaseSelectorPhaseList: string;
  phaseSelectorFavorites: Record<string, string[]>;
  selectedWaveforms: ChannelSegmentTypes.ChannelSegmentDescriptor[];
}
