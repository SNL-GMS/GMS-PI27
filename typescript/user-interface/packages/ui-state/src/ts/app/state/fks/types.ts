/**
 * Different filters that are available
 */

import type { FkTypes } from '@gms/common-model';

import type { ChannelFilterRecord } from '../../../types';

export enum FkThumbnailsFilterType {
  KEYACTIVITYPHASES = 'Key Activity Phases',
  ALL = 'All',
  OPENEVENT = 'Open Event',
  NEEDSREVIEW = 'Needs review'
}

export interface FksState {
  sdIdsToShowFk: string[];
  displayedSDMeasuredValues: FkTypes.FkMeasuredValues | undefined;
  fkPlotsExpandToolbar: boolean;
  currentFkThumbnailFilter: FkThumbnailsFilterType;
  fkChannelFilters: Record<string, ChannelFilterRecord>;
  displayedSignalDetectionId: string;
}
