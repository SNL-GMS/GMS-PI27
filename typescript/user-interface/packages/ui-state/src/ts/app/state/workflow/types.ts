import type { CommonTypes, WorkflowTypes } from '@gms/common-model';
import type { Nullable } from '@gms/common-model/lib/type-util/type-util';

export interface WorkflowState {
  timeRange: Nullable<CommonTypes.TimeRange>;
  stationGroup: WorkflowTypes.StationGroup;
  openIntervalName: string; // e.x AL1
  openActivityNames: string[]; // e.x Event Review
  analysisMode: WorkflowTypes.AnalysisMode | null;
}
