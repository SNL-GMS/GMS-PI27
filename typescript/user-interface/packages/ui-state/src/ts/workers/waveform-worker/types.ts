import type { ChannelTypes, StationTypes, WaveformTypes } from '@gms/common-model';
import type { BeamDefinition } from '@gms/common-model/lib/beamforming-templates/types';
import type { ProcessingMask, TimeRangesByChannel } from '@gms/common-model/lib/channel-segment';
import type { TimeRange } from '@gms/common-model/lib/common';
import type { AxiosRequestConfig } from 'axios';

import type { GetChannelSegmentsByChannelQueryArgs } from '../../app/api/data/waveform/get-channel-segments-by-channel';
import type { UiChannelSegment } from '../../types';

/**
 * The colors to use when building the weavess channel segments
 */
export interface ChannelSegmentColorOptions {
  waveformColor: string;
  labelTextColor: string;
}

export interface AmplitudeBounds {
  amplitudeMax: number;
  amplitudeMin: number;
  amplitudeTotal?: number;
  totalSamplesCount?: number;
  amplitudeMaxSecs?: number;
  amplitudeMinSecs?: number;
}

/**
 * An axios request known to have a waveform query request in its data.
 */
export type WaveformAxiosRequestConfig = AxiosRequestConfig & {
  data: GetChannelSegmentsByChannelQueryArgs;
};

// Props for the ui-beam-processor maskAndBeamWaveforms operation
export interface MaskAndBeamWaveformWorkerProps {
  beamDefinition: BeamDefinition;
  beamStartTime: number;
  beamEndTime: number;
  station: StationTypes.Station;
  filteredChannelSegments: UiChannelSegment<WaveformTypes.Waveform>[];
  processingMasksByChannel: {
    channel: ChannelTypes.Channel;
    processingMasks: ProcessingMask[];
  }[];
  currentInterval: TimeRange;
  missingInputChannels: TimeRangesByChannel[];
}

export interface MaskAndBeamWaveformResult {
  channel: ChannelTypes.Channel;
  uiChannelSegment: UiChannelSegment<WaveformTypes.Waveform>;
}
