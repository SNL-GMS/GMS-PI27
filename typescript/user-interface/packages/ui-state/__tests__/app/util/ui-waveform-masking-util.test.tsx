import {
  PD01Channel,
  processingMaskDefinition,
  qcSegment,
  qcSegment2Version,
  qcSegment3Version2,
  qcSegment4Version,
  qcSegment5Version,
  qcSegment6Version,
  qcSegmentAkasgBHE,
  qcSegmentAkasgBHN,
  qcSegmentVersion
} from '@gms/common-model/__tests__/__data__';
import { ProcessingOperation } from '@gms/common-model/lib/channel-segment/types';
import type { Channel } from '@gms/common-model/lib/station-definitions/channel-definitions/channel-definitions';
import type { WithRequired } from '@gms/common-model/lib/type-util/type-util';
import { MILLISECONDS_IN_SECOND } from '@gms/common-util';
import type { PriorityRequestConfig } from '@gms/ui-workers';
import { renderHook } from '@testing-library/react-hooks';
import type { AxiosResponse } from 'axios';
import Axios from 'axios';
import React from 'react';
import { Provider } from 'react-redux';
import type { AnyAction } from 'redux';
import type { MockStoreCreator } from 'redux-mock-store';
import createMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import type { GetChannelsByNamesTimeRangeQueryArgs } from '../../../src/ts/app';
import type { AppState } from '../../../src/ts/app/store';
import {
  createProcessingMasksFromQCSegmentVersions,
  useCreateProcessingMasksFromChannelSegment
} from '../../../src/ts/app/util/ui-waveform-masking-util';
import { unfilteredSamplesUiChannelSegment } from '../../__data__';
import { appState } from '../../test-util';

const MOCK_UUID = 123456789;

jest.mock('@gms/common-util', () => {
  const actual = jest.requireActual('@gms/common-util');
  return {
    ...actual,
    epochSecondsNow: () => 100,
    uuid4: () => MOCK_UUID
  };
});

jest.mock('../../../src/ts/app/hooks/waveform-hooks', () => {
  const actual = jest.requireActual('../../../src/ts/app/hooks/waveform-hooks');
  return {
    ...actual,
    useViewableInterval: () => [{ startTimeSecs: 0, endTimeSecs: 1000 }, jest.fn()]
  };
});

jest.mock('../../../src/ts/workers/api/fetch-channels-by-names-timerange', () => {
  const actual = jest.requireActual(
    '../../../src/ts/workers/api/fetch-channels-by-names-timerange'
  );
  return {
    ...actual,
    fetchChannelsByNamesTimeRange: async (
      requestConfig: WithRequired<
        PriorityRequestConfig<GetChannelsByNamesTimeRangeQueryArgs>,
        'data'
      >
    ): Promise<Channel[]> => {
      const { endTime, startTime } = requestConfig.data;
      if (startTime == null || endTime == null) {
        throw new Error('Error building mock Channel. startTime and endTime must be defined');
      }

      return Promise.resolve([PD01Channel]);
    }
  };
});

jest.mock('../../../src/ts/app/hooks/operational-time-period-configuration-hooks', () => {
  const app = jest.requireActual(
    '../../../src/ts/app/hooks/operational-time-period-configuration-hooks'
  );
  return {
    ...app,
    useEffectiveTime: jest.fn(() => PD01Channel.effectiveAt + 1000)
  };
});

const MOCK_TIME = 1606818240000;
global.Date.now = jest.fn(() => MOCK_TIME);

const response: AxiosResponse<unknown> = {
  status: 200,
  config: {},
  headers: {},
  statusText: '',
  data: {}
};

const mockAxiosRequest = jest.fn().mockImplementation(async requestConfig => {
  const { url } = requestConfig;
  if (url.includes('processing-mask')) {
    return Promise.resolve({
      ...response,
      data: {
        processingMaskDefinitionByPhaseByChannel: [
          {
            channel: requestConfig.data.channels[0],
            processingMaskDefinitionByPhase: {
              [requestConfig.data.phaseTypes[0]]: {
                ...processingMaskDefinition,
                processingOperation: ProcessingOperation.ROTATION
              }
            }
          }
        ]
      }
    });
  }
  if (url.includes('qc-segment')) {
    if (requestConfig.data.channels.find(chan => chan.name.includes('BHE'))) {
      return Promise.resolve({
        ...response,
        data: [
          {
            ...qcSegmentAkasgBHE,
            versionHistory: [
              {
                ...qcSegmentAkasgBHE.versionHistory[0],
                startTime: requestConfig.data.startTime,
                endTime: requestConfig.data.endTime
              }
            ]
          }
        ]
      });
    }
    if (requestConfig.data.channels.find(chan => chan.name.includes('BHN'))) {
      return Promise.resolve({
        ...response,
        data: [qcSegmentAkasgBHN]
      });
    }
  }

  const success = 'success';
  return Promise.resolve(success);
});
Axios.request = mockAxiosRequest;

describe('ui waveform masking util', () => {
  describe('createProcessingMasksFromQCSegmentVersions', () => {
    const segmentVersions = [
      qcSegmentVersion,
      qcSegment2Version,
      qcSegment3Version2,
      qcSegment4Version,
      qcSegment5Version
    ];

    it('is defined', () => {
      expect(createProcessingMasksFromQCSegmentVersions).toBeDefined();
    });
    it('filters out segments that do not match the mask definition', () => {
      const processingMasks = createProcessingMasksFromQCSegmentVersions(
        segmentVersions,
        processingMaskDefinition
      );
      expect(processingMasks[0].maskedQcSegmentVersions).toEqual([
        qcSegment3Version2,
        qcSegment5Version
      ]);
    });
    it('groups segments based on the threshold and sets teh start and end time to the min and max', () => {
      const processingMasks = createProcessingMasksFromQCSegmentVersions(
        segmentVersions.concat(qcSegment6Version),
        processingMaskDefinition
      );

      expect(processingMasks[0].maskedQcSegmentVersions).toEqual([
        qcSegment3Version2,
        qcSegment5Version
      ]);
      expect(processingMasks[0].startTime).toEqual(qcSegment3Version2.startTime);
      expect(processingMasks[0].endTime).toEqual(qcSegment5Version.endTime);
      expect(processingMasks[1].maskedQcSegmentVersions).toEqual([qcSegment6Version]);
      expect(processingMasks[1].startTime).toEqual(qcSegment6Version.startTime);
      expect(processingMasks[1].endTime).toEqual(qcSegment6Version.endTime);
    });
    it('generates a new UUID for the processing mask', () => {
      const processingMasks = createProcessingMasksFromQCSegmentVersions(
        segmentVersions.concat(qcSegment6Version),
        processingMaskDefinition
      );

      expect(processingMasks[0].id).toEqual(MOCK_UUID);
    });
    it('sets the effective at to now in seconds', () => {
      const processingMasks = createProcessingMasksFromQCSegmentVersions(
        segmentVersions.concat(qcSegment6Version),
        processingMaskDefinition
      );

      expect(processingMasks[0].effectiveAt).toEqual(Date.now() / MILLISECONDS_IN_SECOND);
    });

    it('sets the processing operation to that of the mask definition', () => {
      const processingMasks = createProcessingMasksFromQCSegmentVersions(
        segmentVersions.concat(qcSegment6Version),
        processingMaskDefinition
      );

      expect(processingMasks[0].processingOperation).toEqual(
        processingMaskDefinition.processingOperation
      );
    });
  });

  describe('useCreateProcessingMasksFromChannelSegment', () => {
    it('creates processing masks', async () => {
      const mockStoreCreator: MockStoreCreator<AppState, AnyAction> = createMockStore([thunk]);

      const store = mockStoreCreator({
        ...appState,
        data: {
          ...appState.data,
          channels: { ...appState.data.channels, raw: { [PD01Channel.name]: PD01Channel } },
          qcSegments: { [PD01Channel.name]: { [qcSegment.id]: qcSegment } },
          processingMaskDefinitionsByChannels: [
            {
              channel: PD01Channel,
              processingMaskDefinitions: { P: { EVENT_BEAM: processingMaskDefinition } }
            }
          ]
        }
      });

      const { result } = renderHook(() => useCreateProcessingMasksFromChannelSegment(), {
        wrapper: (props: React.PropsWithChildren<unknown>) => (
          <Provider store={store}>{props.children}</Provider>
        )
      });
      const returnValue = await result.current(
        unfilteredSamplesUiChannelSegment,
        ProcessingOperation.EVENT_BEAM,
        'P'
      );
      expect(returnValue).toMatchSnapshot();
    });

    it('creates returns the original channel if no masks are applied', async () => {
      const mockStoreCreator: MockStoreCreator<AppState, AnyAction> = createMockStore([thunk]);

      const store = mockStoreCreator({
        ...appState,
        data: {
          ...appState.data,
          channels: { ...appState.data.channels, raw: { [PD01Channel.name]: PD01Channel } },
          qcSegments: { [PD01Channel.name]: {} },
          processingMaskDefinitionsByChannels: [
            {
              channel: PD01Channel,
              processingMaskDefinitions: { P: { EVENT_BEAM: processingMaskDefinition } }
            }
          ]
        }
      });

      const { result } = renderHook(() => useCreateProcessingMasksFromChannelSegment(), {
        wrapper: (props: React.PropsWithChildren<unknown>) => (
          <Provider store={store}>{props.children}</Provider>
        )
      });
      const returnValue = await result.current(
        unfilteredSamplesUiChannelSegment,
        ProcessingOperation.EVENT_BEAM,
        'P'
      );
      expect(returnValue.channel).toEqual(PD01Channel);
    });
  });
});
