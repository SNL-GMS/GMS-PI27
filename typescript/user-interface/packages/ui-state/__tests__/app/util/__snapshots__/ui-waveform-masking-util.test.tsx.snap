// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui waveform masking util useCreateProcessingMasksFromChannelSegment creates processing masks 1`] = `
{
  "channel": {
    "canonicalName": "PDAR.PD01.SHZ/masked/0501421b22b559405b02a551b38508334a9c51af6ea3c20b524f93a7501eb9f4",
    "channelBandType": "SHORT_PERIOD",
    "channelDataType": "SEISMIC",
    "channelInstrumentType": "HIGH_GAIN_SEISMOMETER",
    "channelOrientationCode": "Z",
    "channelOrientationType": "VERTICAL",
    "configuredInputs": [
      {
        "effectiveAt": 1636503404,
        "name": "PDAR.PD01.SHZ",
      },
    ],
    "description": "Raw Channel created from ReferenceChannel 2fabc2d3-858b-3e85-9f47-e2ee72060f0b with version d767395c-850e-36f8-a6f2-a1c7398440e4 Masked samples removed.",
    "effectiveAt": 1636503404,
    "effectiveForRequestTime": 1636503405,
    "effectiveUntil": 1660701599.984,
    "location": {
      "depthKm": 0.0381,
      "elevationKm": 2.192,
      "latitudeDegrees": 42.7765,
      "longitudeDegrees": -109.58314,
    },
    "name": "PDAR.PD01.SHZ/masked/0501421b22b559405b02a551b38508334a9c51af6ea3c20b524f93a7501eb9f4",
    "nominalSampleRateHz": 20,
    "orientationAngles": {
      "horizontalAngleDeg": -1,
      "verticalAngleDeg": 0,
    },
    "processingDefinition": {
      "appliedQcSegmentCategoryAndTypes": [
        {
          "category": "ANALYST_DEFINED",
          "type": "AGGREGATE",
        },
      ],
      "maskedSegmentMergeThreshold": 50,
      "processingOperation": "EVENT_BEAM",
      "taperDefinition": {
        "taperFunction": "COSINE",
        "taperLengthSamples": 100,
      },
    },
    "processingMetadata": {
      "CHANNEL_GROUP": "PD01",
    },
    "response": undefined,
    "station": {
      "effectiveAt": undefined,
      "name": "PDAR",
    },
    "units": "NANOMETERS_PER_COUNT",
  },
  "processingMasks": [
    {
      "appliedToRawChannel": {
        "name": "PDAR.PD01.SHZ",
      },
      "effectiveAt": 1606818240,
      "endTime": 1636503704,
      "id": 123456789,
      "maskedQcSegmentVersions": [
        {
          "category": "ANALYST_DEFINED",
          "channels": [
            {
              "effectiveAt": 1636503404,
              "name": "PDAR.PD01.SHZ",
            },
          ],
          "createdBy": "User 1",
          "discoveredOn": undefined,
          "endTime": 1636503704,
          "id": {
            "effectiveAt": 0,
            "parentQcSegmentId": "qcSegmentTestId",
          },
          "rationale": "",
          "rejected": false,
          "stageId": undefined,
          "startTime": 1636503404,
          "type": "AGGREGATE",
        },
      ],
      "processingOperation": "EVENT_BEAM",
      "startTime": 1636503404,
    },
  ],
}
`;
